### PR TITLE
[PromotionCoupon] Non reusable coupons after cancelling the orders

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | 
| License         | MIT

Zrzut ekranu 2019-04-16 o 10 18 25

### CODE REVIEW

File: src/Sylius/Component/Core/spec/Model/PromotionCouponSpec.php [hunk 1]
```diff
@@ -33,4 +33,15 @@ public function its_per_customer_usage_limit_should_be_mutable()
         $this->setPerCustomerUsageLimit(10);
         $this->getPerCustomerUsageLimit()->shouldReturn(10);
     }
+
+    public function its_reusable_from_cancelled_orders_flag_is_true_by_default()
```
Comment (Tomanhez) [inline src/Sylius/Component/Core/spec/Model/PromotionCouponSpec.php]: missing type : void

File: src/Sylius/Component/Core/spec/Model/PromotionCouponSpec.php [hunk 2]
```diff
@@ -33,4 +33,15 @@ public function its_per_customer_usage_limit_should_be_mutable()
...
+    {
+        $this->isReusableFromCancelledOrders()->shouldReturn(true);
+    }
+
+    public function its_reusable_from_cancelled_orders_flag_is_mutable()
```
Comment (Tomanhez) [inline src/Sylius/Component/Core/spec/Model/PromotionCouponSpec.php]: and here

File: features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_per_customer_usage_limit.feature [hunk 3]
```diff
@@ -35,28 +35,38 @@ Feature: Applying promotion coupon with per customer usage limit
...
         When I add product "PHP T-Shirt" to the cart
         And I use coupon with code "SANTA2016"
         Then I should be notified that the coupon is invalid
         And my cart total should be "$100.00"
         And there should be no discount
 
     @ui
-    Scenario: Cancelled orders should not influence per customer usage limit
-        Given I placed an order "#00000022"
+    Scenario: Cancelled orders are not affecting per customer usage limit by default
```
Comment (bartoszpietrzak1994) [inline features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_per_customer_usage_limit.feature]: What about `Cancelled orders do not affect per customer usage limit by default` ?

File: features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_per_customer_usage_limit.feature [hunk 4]
```diff
@@ -35,28 +35,38 @@ Feature: Applying promotion coupon with per customer usage limit
...
-        And I bought a single "PHP Socks" using "SANTA2016" coupon
-        And I chose "Free" shipping method to "United States" with "Cash on Delivery" payment
         But I cancelled this order
         When I add product "PHP T-Shirt" to the cart
         And I use coupon with code "SANTA2016"
         Then my cart total should be "$90.00"
         And my discount should be "-$10.00"
+
+    @ui
+    Scenario: Cancelled orders are affecting per customer usage limit
```
Comment (bartoszpietrzak1994) [inline features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_per_customer_usage_limit.feature]: What about `Cancelled orders affect per customer usage limit` ?

File: features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_usage_limit.feature [hunk 5]
```diff
@@ -26,3 +29,29 @@ Feature: Applying promotion coupon with usage limit
         Then I should be notified that the coupon is invalid
         And my cart total should be "$100.00"
         And there should be no discount
+
+    @ui
+    Scenario: Cancelled orders are not affecting the usage limit by default
```
Comment (bartoszpietrzak1994) [inline features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_usage_limit.feature]: What about `Cancelled orders do not affect usage limit by default` ?

File: features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_usage_limit.feature [hunk 6]
```diff
@@ -26,3 +29,29 @@ Feature: Applying promotion coupon with usage limit
...
+        And I bought a single "PHP T-Shirt" using "SANTA2016" coupon
+        And I chose "Free" shipping method to "United States" with "Cash on Delivery" payment
+        But I cancelled this order
+        When I add product "PHP T-Shirt" to the cart
+        And I use coupon with code "SANTA2016"
+        Then my cart total should be "$90.00"
+        And my discount should be "-$10.00"
+
+    @ui
+    Scenario: Cancelled orders are affecting usage limit
```
Comment (bartoszpietrzak1994) [inline features/promotion/applying_promotion_coupon/applying_promotion_coupon_with_usage_limit.feature]: What about `Cancelled orders affect usage limit` ?

File: src/Sylius/Component/Core/Promotion/Modifier/OrderPromotionsUsageModifier.php [hunk 7]
```diff
@@ -41,8 +42,19 @@ public function decrement(OrderInterface $order): void
...
         }
 
+        /** @var PromotionCouponInterface|null $promotionCoupon */
         $promotionCoupon = $order->getPromotionCoupon();
-        if (null !== $promotionCoupon) {
+        if (null === $promotionCoupon) {
+            return;
+        }
+
+        if (OrderInterface::STATE_CANCELLED !== $order->getState()) {
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Promotion/Modifier/OrderPromotionsUsageModifier.php]: Why don't we merge both if statements?
```suggestion
        if (OrderInterface::STATE_CANCELLED !== $order->getState() || $promotionCoupon->isReusableFromCancelledOrders()) {
```

**Comment by CoderMaggie** [comment]
@lchrusciel As a PO, I wanted it to be on coupon level because I've assumed this feature only for coupon based promotions, and thus configurable per each of the coupons.

**Comment by lchrusciel** [comment]
Thank you, Grzegorz! :tada:

**Pull request merged**